### PR TITLE
[QA] Add tests for JS files

### PR DESCRIPTION
**Pull request by vladimir-mck**
## Description

Closes [#1927](https://github.com/McK-Internal/vizro-internal/issues/1927) 
Closes  [#1975](https://github.com/McK-Internal/vizro-internal/issues/1975)

- Introduces tests for JavaScript files using Jest.
- To execute, run the following command from  vizro-core `hatch run test-js`
- Adds workflow for js tests

## Notice

- [ ] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

    - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
    - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
    - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
    - I have not referenced individuals, products or companies in any commits, directly or indirectly.
    - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/src/vizro/static/js/models/range_slider.js [hunk 1]
```diff
@@ -13,7 +26,19 @@ function update_range_slider_values(
...
 
   // text form component is triggered (start_value or end_value)
-  if (isNaN(start_value) || isNaN(end_value) || start_value > end_value) {
+  // Check if end_value or start_value are a valid number, JavaScript's Number.isNaN is used to check for NaN.
+  // If start_value or end_value are not a valid number, throw PreventUpdate to stop further processing.
+  // This includes cases where end_value is null, or an empty string.
+  // JavaScript's Number() will convert empty strings and null to 0, so we check for that explicitly.
+  if (
+    isNaN(start_value) ||
+    start_value === null ||
```
Comment (pruthvip15) [inline vizro-core/src/vizro/static/js/models/range_slider.js:35]: nit: 
!start_value and !end_value might be a simpler check, and it will check for start_value === undefined also.

Comment (pruthvip15) [reply vizro-core/src/vizro/static/js/models/range_slider.js:35]: i missed the zero bit.

this might help

```
function isFalsyButNotZero(value) {
  return !value && value !== 0;
}

// Examples:
isFalsyButNotZero(0);        // false
isFalsyButNotZero("");       // true
isFalsyButNotZero(null);     // true
isFalsyButNotZero(undefined); // true
isFalsyButNotZero(false);    // true
isFalsyButNotZero(NaN);      // true
isFalsyButNotZero(1);        // false
isFalsyButNotZero("hello");  // false
```

File: vizro-core/src/vizro/static/js/models/slider.js [hunk 2]
```diff
@@ -8,7 +17,15 @@ function update_slider_values(slider_value, end_value, slider_component_id) {
...
 
   // text form component is triggered (end_value)
-  if (isNaN(end_value)) {
+  // Check if end_value is a valid number, JavaScript's Number.isNaN is used to check for NaN.
+  // If end_value is not a valid number, throw PreventUpdate to stop further processing.
+  // This includes cases where end_value is null, or an empty string.
+  // JavaScript's Number() will convert empty strings and null to 0, so we check for that explicitly.
+  if (
+    end_value === null ||
+    end_value === "" ||
```
Comment (pruthvip15) [inline vizro-core/src/vizro/static/js/models/slider.js:26]: !

File: vizro-core/src/vizro/static/js/models/slider.js [hunk 3]
```diff
@@ -1,3 +1,12 @@
+/**
+ * Updates slider values based on user interaction with either the slider component or text input.
+ *
+ * @param {number} slider_value - The current value of the slider component
+ * @param {number} end_value - The value from the text input component
+ * @param {string} slider_component_id - The ID of the slider component to identify trigger source
+ * @returns {Array} Array containing [text_input_value, slider_value] where one element is updated and the other is dash_clientside.no_update
+ * @throws {dash_clientside.PreventUpdate} When end_value is NaN (invalid number input)
```
Comment (pruthvip15) [inline vizro-core/src/vizro/static/js/models/slider.js:8]: Unrelated suggestion. after seeing the code, i feel the js doesnt compressed when the final dashboard is created. compressing the js, makes the startup time of the app a little faster.  we can look into it

File: vizro-core/src/vizro/static/js/models/range_slider.js (hunk 1 above)
Comment (vladimir-mck) [reply vizro-core/src/vizro/static/js/models/range_slider.js:35]: No,  0 will also be a false value and in this case 0 is an acceptable value.  `isNaN` returns false for null and empty strings so need to check for those as well.

File: vizro-core/src/vizro/static/js/models/slider.js (hunk 3 above)
Comment (pruthvip15) [reply vizro-core/src/vizro/static/js/models/slider.js:8]: fyi @huong-li-nguyen

File: vizro-core/tests/js/models/dashboard.test.js [hunk 4]
```diff
@@ -1,41 +1,540 @@
...
+
+describe("dashboard.js functions", () => {
+  let setAttributeSpy;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    // Reset window dimensions to default desktop size
+    global.window.innerWidth = 1024;
+    global.window.innerHeight = 768;
```
Comment (pruthvip15) [inline vizro-core/tests/js/models/dashboard.test.js]: nit: optional: we can keep both contatnts in a object, and use it at both places line 24 and here

File: vizro-core/tests/js/models/dropdown.test.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+      expect(result[0]).toBe(true);
+      expect(result[1]).toEqual(realOptions);
+    });
+  });
+
+  describe("when regular dropdown item is clicked", () => {
+    test("should set select all to true when all real options are selected", () => {
+      const result = window.dash_clientside.dropdown.update_dropdown_select_all(
+        ["option1", "option2", "option3"],
```
Comment (pruthvip15) [inline vizro-core/tests/js/models/dropdown.test.js]: nit optinal, can use loops to make it easier, if you find that wil be useful

```
describe("update_dropdown_select_all", () => {
  const mockOptions = [
    { value: "option1", label: "Option 1" },
    { value: "option2", label: "Option 2" },
    { value: "option3", label: "Option 3" }
  ];

  const updateDropdownSelectAll = (selectedValues) => {
    return window.dash_clientside.dropdown.update_dropdown_select_all(
      selectedValues,
      mockOptions
    );
  };

  describe("when all real options are selected", () => {
    it("should set select all to true", () => {
      const selectedValues = ["option1", "option2", "option3"];
      const [selectAllValue, updateValue] = updateDropdownSelectAll(selectedValues);

      expect(selectAllValue).toBe(true);
      expect(updateValue).toBe(global.dash_clientside.no_update);
    });
  });

  describe("when not all real options are selected", () => {
    it.each([
      {
        description: "two options selected",
        selectedValues: ["option1", "option2"],
        expectedSelectAll: false
      },
      {
        description: "one option selected", 
        selectedValues: ["option1"],
        expectedSelectAll: false
      },
      {
        description: "no options selected",
        selectedValues: [],
        expectedSelectAll: false
      }
    ])("should set select all to $expectedSelectAll when $description", ({ selectedValues, expectedSelectAll }) => {
      const [selectAllValue, updateValue] = updateDropdownSelectAll(selectedValues);

      expect(selectAllValue).toBe(expectedSelectAll);
      expect(updateValue).toBe(global.dash_clientside.no_update);
    });
  });

  // Alternative approach using test.each if you prefer more explicit test cases
  describe("edge cases", () => {
    it.each([
      ["option1", false],
      ["option1", "option2", false], 
      ["option1", "option2", "option3", true],
      [[], false]
    ])("should return correct select all state for selected values: %p", (...selectedValues) => {
      const expectedSelectAll = selectedValues.length === mockOptions.length;
      const [selectAllValue, updateValue] = updateDropdownSelectAll(selectedValues);

      expect(selectAllValue).toBe(expectedSelectAll);
      expect(updateValue).toBe(global.dash_clientside.no_update);
    });
  });
});
```

**Review by pruthvip15 — APPROVED** [review]
there is one check with undefined. just double check that

rest all look good to me

**Review by huong-li-nguyen — APPROVED** [review]
Feel free to merge after incorporating @pruthvip15 feedback if it makes sense to you 👍
